sdks/java/core/src/main/java/org/apache/beam/sdk/io/FileBasedSink.java

```diff
@@ -340,36 +367,44 @@ public ResourceId apply(ResourceId input) {
   }
 
   /**
-   * Construct a {@link FileBasedSink} with the given filename policy and output channel type.
+   * Construct a {@link FileBasedSink} with the given temp directory, producing uncompressed
+   * files.
+   */
+  @Experimental(Kind.FILESYSTEM)
+  public FileBasedSink(ValueProvider<ResourceId> tempDirectoryProvider,
+                       DynamicDestinations<?, DestinationT> dynamicDestinations) {
+    this(tempDirectoryProvider, dynamicDestinations, CompressionType.UNCOMPRESSED);
+
+  }
+
+  /**
+   * Construct a {@link FileBasedSink} with the given temp directory and output channel type.
    */
   @Experimental(Kind.FILESYSTEM)
   public FileBasedSink(
-      ValueProvider<ResourceId> baseOutputDirectoryProvider,
-      FilenamePolicy filenamePolicy,
+      ValueProvider<ResourceId> tempDirectoryProvider,
+      DynamicDestinations<?, DestinationT> dynamicDestinations,
       WritableByteChannelFactory writableByteChannelFactory) {
-    this.baseOutputDirectoryProvider =
-        NestedValueProvider.of(baseOutputDirectoryProvider, new ExtractDirectory());
-    this.filenamePolicy = filenamePolicy;
+    this.tempDirectoryProvider =
+        NestedValueProvider.of(tempDirectoryProvider, new ExtractDirectory());
+    this.dynamicDestinations = checkNotNull(dynamicDestinations);
     this.writableByteChannelFactory = writableByteChannelFactory;
   }
 
   /**
-   * Returns the base directory inside which files will be written according to the configured
-   * {@link FilenamePolicy}.
+   * Return the {@link DynamicDestinations} used.
    */
-  @Experimental(Kind.FILESYSTEM)
-  public ValueProvider<ResourceId> getBaseOutputDirectoryProvider() {
-    return baseOutputDirectoryProvider;
+  public <UserT> DynamicDestinations<UserT, DestinationT> getDynamicDestinations() {
```

Comment:
There is: `public abstract class FileBasedSink<T, DestinationT>`, `class WriteOperation<T, DestinationT>` etc.